### ISSUE TITLE
.NET Core Compatibility

### ISSUE THREAD

**Issue by peske** [zeromq/clrzmq4#71]
You're probably aware that .NET Core becomes very important target these days, and that it is the future when it comes to portability. Do you have any plans to support it? +1 for this.

Thanks!

**Comment by metadings** [comment]
Yes, there should be compiler directive NETCORE, so you can build this against .NET Core.

**Comment by peske** [comment]
I don't understand your answer. You are suggesting that it is possible now or that it'll be implemented in the future?

**Comment by metadings** [comment]
No it is not possible now. It will be implemented in the future.

**Comment by aerotog** [comment]
The necessary API were added in .Net Standard 2.0. Changing the project's target to .Net Standard 2.0 will [enable support for .Net Core 2.0](https://github.com/dotnet/standard/blob/master/docs/versions.md). The catch is the current code targets .Net Framework 4.0. By changing the target to Standard 2.0, you change the compatible frameworks to .Net Framework 4.6.1 and Mono 5.4.

.Net Framework 4.6.1 is almost 2 years old so requiring it shouldn't be a big deal. Mono 5.4 went stable 2 weeks ago so there could be compatibility problems for people not staying up to date. That said, making this compatible with .Net Core 2.0 is a big win and I think a good step forward.

Once the pull request from #137 is merged, I will be opening another that changes the target for the project to .Net Standard 2.0. I have tested it .Net Framework 4.6.1 and .Net Core 2.0 on Windows and Linux and it works as expected.

**Comment by aerotog** [comment]
I should finally have time to port this to .Net Standard 2.0 in the next couple weeks.

While the change from #137 was reverted, I should still be able to update the target to Standard 2.0 with few code changes. The hardest part will likely be getting it to pass testing and CI as building and testing .Net Standard projects is different from .Net Framework projects. (Mainly using 'dotnet' vs 'msbuild')

I would consider this a significant breaking change. This means we will need to consider whether to rev the major/minor version.

**Comment by aerotog** [comment]
I'm very tempted to make some severe changes to the project. (i.e. removing the Mono project) It seems if there is a single .Net Standard project to build, it would be able to serve all use cases. (Except for people targeting older Mono/Framework versions)

The main problem I see is since this uses Mono specific code for loading the libzmq library, I'll need to figure out alternatives for each platform. (Particularly for Mac OSX for which I don't have a test environment) This is related to the change I tried for #137.

All this is high risk as it could easily break down stream consumers.

I'll be spending some time looking at the best option and open a PR to get feedback when it's ready.

**Comment by Cvanvollenstee** [comment]
Any news regarding the .net core implementation; we are currently using gRpc and want to move over to zerorpc but this .net core compatibility issues is stopping us at the moment. So any feedback would be much appreciated.

**Comment by aerotog** [comment]
I started porting clrzmq4 to .Net Standard and got it working with .Net Core on Windows with very few changes. My Linux work has been rescheduled, so it will be another month or two before I get time to work on that.

**Comment by Cvanvollenstee** [comment]
Awesome!!! We are very exited here on our end and awaiting the go ahead to start testing your work.

**Comment by tjaart** [comment]
What still needs to be done to get it working for .net standard on Linux @aerotog ?

**Comment by aerotog** [comment]
@tjaart, the technical work to get this project working in .Net Standard on Linux is not difficult. I did it a couple months ago as a quick and dirty proof of concept.

The hard part is deciding how to incorporate the changes into the main project. Moving to .Net Standard and/or changing the existing Posix platform will break existing Mono users and require later versions of both .Net Framework and Mono than are currently specified.

When I eventually return to this, my plan is to implement my "ideal" solution in a fork where I use a single .Net Standard project and remove the two existing projects (one targeting Mono and one targeting .Net Framework). I will also remove the Mono references and replace them with libdl reference like I tried before [here](https://github.com/aerotog/clrzmq4/commit/ed0123d142e37d9a72d81854fcbb579f83a1fa33). That change will enable operation on Linux.

**Comment by aerotog** [comment]
Looks like the source of the Linux problem I was talking about was removed [here](https://github.com/zeromq/clrzmq4/pull/142
). I was working with an older version that had dll import conflicts becauce of the __Internal references. Hopefully this means a Standard build would work in Core as is. I'll try to test it by next week.

I've opened a PR to demonstrate the extent of changes required to move to .Net Standard. We could theoretically keep the old .Net Framework and Mono solution and project files to maintain backwards compatibility, but I think it would be nice to have a clean cut and move over to .Net Standard alone.

My local branch was based on an older version of this project, so I haven't tested this PR aside from running the tests in the ZeroMQTest project. Most of them pass.

I don't expect this PR to be approved, but please use it as a reference to see how we could move the project to Standard.

**Comment by aerotog** [comment]
I've tested a local build of the .Net Standard project from my PR. With the changes from PR #142, it works in .Net Core apps on both Windows and Linux. If the maintainers agree on a path for migrating to .Net Standard, I'd be more than happy to help fix the PR and get it passing the CI checks.

**Comment by MendelMonteiro** [comment]
Any news from the maintainers?  @metadings @sigiesec

**Comment by AntonioGMuriana** [comment]
I see that PR #142 is already merged. Could it be compiled in .Net Core?

**Comment by MendelMonteiro** [comment]
@AntonioGMuriana this project seems pretty dead - there have been no new commits since the end of last year and very little interaction on issues/pull requests

**Comment by Cvanvollenstee** [comment]
It is sad this project had a great future.

**Comment by AntonioGMuriana** [comment]
I managed to compile to .Net Core 2 with the project file from pull #154 and removing unneeded "using" from ZSymbol.cs.

**Comment by aerotog** [comment]
There were complications with the MacOS build which I didn't have an easy way of testing. Also the move to .Net Standard was rejected due to version incompatibility with old Mono and .Net Framework versions.

The code from PR #154 should compile like @AntonioGMuriana said. Feel free to fork it and give the .Net Standard version a shot. I had it working locally. Due to various concerns including lack of .Net Core compatibility, my team has decided to move away from ZMQ. The final replacement is undecided but we're currently trying RabbitMQ.

**Comment by metadings** [comment]
Now please look into the one of the latest commits #168 : I added new files `clrzmq4.netcore.sln`, `ZeroMQ.netcore.csproj` and `ZGuideExamples/ZGuideExamples.netcore.csproj`.
You can open those files using Visual Studio, Visual Studio Code and MonoDevelop or your favorite editor, this is going to build on .NET Standard 2.0.

Please also look into the new `README.md` how I'm running `ZGuide.dll` implicitly, using `dotnet` do
```
nuget restore clrzmq4.netcore.sln
msbuild clrzmq4.netcore.sln
dotnet run --project ZGuideExamples/ZGuideExamples.netcore.csproj Espresso
```
